flink-connectors/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/committer/PulsarCommitter.java

```diff
@@ -0,0 +1,167 @@
+ */
+@Internal
+public class PulsarCommitter implements Committer<PulsarCommittable>, Closeable {
+    private static final Logger LOG = LoggerFactory.getLogger(PulsarCommitter.class);
+
+    private final SinkConfiguration sinkConfiguration;
+
+    private PulsarClient pulsarClient;
+    private TransactionCoordinatorClient coordinatorClient;
+
+    public PulsarCommitter(SinkConfiguration sinkConfiguration) {
+        this.sinkConfiguration = sinkConfiguration;
+    }
+
+    @Override
+    public void commit(Collection<CommitRequest<PulsarCommittable>> requests)
+            throws IOException, InterruptedException {
+        TransactionCoordinatorClient client = transactionCoordinatorClient();
+
+        for (CommitRequest<PulsarCommittable> request : requests) {
+            PulsarCommittable committable = request.getCommittable();
+            TxnID txnID = committable.getTxnID();
+            String topic = committable.getTopic();
+
+            LOG.debug("Start committing the Pulsar transaction {} for topic {}", txnID, topic);
+            try {
+                client.commit(txnID);
+            } catch (TransactionCoordinatorClientException e) {
+                // This is a known bug for Pulsar Transaction. We have to use instance of.
+                TransactionCoordinatorClientException ex = PulsarTransactionUtils.unwrap(e);
+                if (ex instanceof CoordinatorNotFoundException) {
+                    LOG.error(
+                            "We couldn't find the Transaction Coordinator from Pulsar broker {}. "
+                                    + "Check your broker configuration.",
+                            committable,
+                            ex);
+                    request.signalFailedWithKnownReason(ex);
+                } else if (ex instanceof InvalidTxnStatusException) {
+                    LOG.error(
+                            "Unable to commit transaction ({}) because it's in an invalid state. "
+                                    + "Most likely the transaction has been aborted for some reason. "
+                                    + "Please check the Pulsar broker logs for more details.",
+                            committable,
+                            ex);
+                    request.signalAlreadyCommitted();
+                } else if (ex instanceof TransactionNotFoundException) {
+                    if (request.getNumberOfRetries() == 0) {
+                        LOG.error(
+                                "Unable to commit transaction ({}) because it's not found on Pulsar broker. "
+                                        + "Most likely the checkpoint interval exceed the transaction timeout.",
+                                committable,
+                                ex);
+                        request.signalFailedWithKnownReason(ex);
+                    } else {
+                        request.signalAlreadyCommitted();
```

Comment:
This case feels strange. Is it really possible that the commit succeeded although the commit failed earlier and was retried.

Reply:
This could happen when the network connecting issue occurs. Pulsar puts the request into a modified protobuf protocol and sends it through netty-based client async.

The connection could timeout but the transaction has been submitted. Pulsar couldn't tell you this situation but returned a transaction not found error instead in this case due to its simplified design. We have to think the transaction has been committed because this isn't the first submission.

The only shortage for this handle is that Pulsar also returns transaction not found exception when the transaction is timeout, which is likely to lead to ambiguity. I have started a discussion with my teammates to see if we can better process this exception.

I would add a warning log to this condition. We can just put this aside in a while.